### PR TITLE
ENH: implement cpython/time.pxd time, localtime

### LINKED ISSUE #3733: ENH: cpython.time to expose stdlib time.localtime etc
There is a `libc.time` available from which I _think_ I can patch together something equivalent to https://github.com/python/cpython/blob/master/Modules/timemodule.c#L490, but itd be nice to have a canonical version exposed directly.

### PR DESCRIPTION
**Pull request by jbrockmendel**
closes #3733

### CODE REVIEW

File: Cython/Includes/cpython/time.pxd [hunk 1]
```diff
@@ -0,0 +1,39 @@
+"""
+Cython implementation of (parts of) the standard library time module.
+"""
+
+from libc.stdint cimport int64_t
+
+ctypedef int64_t _PyTime_t
```
Comment (scoder) [inline Cython/Includes/cpython/time.pxd:7]: This is defined in `pytime.h` (included from `Python.h`), so this should be declared `extern` (together with the functions below).

File: Cython/Includes/cpython/time.pxd [hunk 2]
```diff
@@ -0,0 +1,39 @@
...
+    double _PyTime_AsSecondsDouble(_PyTime_t t) nogil
+
+from libc.time cimport (
+    tm,
+    time_t,
+    localtime as libc_localtime,
+)
+
+
+cpdef inline double time() nogil:
```
Comment (scoder) [inline Cython/Includes/cpython/time.pxd:20]: I think these functions should just be `cdef`. There use is to get cimported and called from Cython code, so there is no need for a Python wrapper. Besides, I actually wonder what `cpdef` does at all inside of a `.pxd` file…

File: tests/run/time_pxd.pyx [hunk 3]
```diff
@@ -0,0 +1,37 @@
...
+    >>> assert ltp.tm_hour == ltc.tm_hour
+    >>> assert ltp.tm_min == ltc.tm_min
+    >>> assert ltp.tm_sec == ltc.tm_sec
+    >>> assert ltp.tm_wday == ltc.tm_wday
+    >>> assert ltp.tm_yday == ltc.tm_yday
+    >>> assert ltp.tm_isdst == ltc.tm_isdst
+    """
+    # Note: the tm_sec assertion may fail in corner cases when the time.localtime
+    #  call is just before the end of a second and the ctime.localtime
+    #  is just after.
```
Comment (scoder) [inline tests/run/time_pxd.pyx:34]: Which is fairly annoying. It may even occur for the year field in one of those corner cases.

Maybe we could mitigate this inside of the test function by testing if the seconds field is identical for both calls, and repeat the first call if not. It seems *very* unlikely that all three calls would return different results in this setup.

File: Cython/Includes/cpython/time.pxd [hunk 4]
```diff
@@ -0,0 +1,39 @@
...
+    tic = _PyTime_GetSystemClock()
+    return _PyTime_AsSecondsDouble(tic)
+
+
+cpdef tm localtime() nogil:
+    """
+    Analogue to the stdlib time.localtime.  The returned struct
+    has some entries that the stdlib version does not: tm_gmtoff, tm_zone
+
+    When called from python, this returns a dict instead of a struct.
```
Comment (scoder) [inline Cython/Includes/cpython/time.pxd:32]: This would happen on the caller side anyway when passing the return value back to Python. No need for a `cpdef` even then.

**Comment by jbrockmendel** [comment]
Updated with suggested edits.

Out of curiosity, why the pattern of putting the assertions in doctests?  Why not just assert in `test_foo`?

**Comment by TeamSpen210** [comment]
The functions are compiled as Cython code, but the doctests are run in regular Python. This way we know they're doing the correct thing, since they'd be unaffected by any Cython bugs.

File: Cython/Includes/cpython/time.pxd [hunk 5]
```diff
@@ -2,11 +2,8 @@
 Cython implementation of (parts of) the standard library time module.
 """
 
-from libc.stdint cimport int64_t
-
-ctypedef int64_t _PyTime_t
-
 cdef extern from "pytime.h":
+    ctypedef _PyTime_t  # alias for int64_t
```
Comment (scoder) [inline Cython/Includes/cpython/time.pxd:6]: This declares it as `object`. :)

You still want to use `int64_t` here to tell Cython what it is, it's just that the `cdef extern` block lets Cython know that the actual declaration is in the header file and not here, and that it should not generate its own C declaration for it.

File: Cython/Includes/cpython/time.pxd [hunk 6]
```diff
@@ -0,0 +1,35 @@
...
+
+cdef inline double time() nogil:
+    cdef:
+        _PyTime_t tic
+
+    tic = _PyTime_GetSystemClock()
+    return _PyTime_AsSecondsDouble(tic)
+
+
+cdef tm localtime() nogil:
```
Comment (scoder) [inline Cython/Includes/cpython/time.pxd:27]: `.pxd` files can only have `inline` functions since they are not modules by themselves. Modules that cimport these functions must `inline` them on their side.
```suggestion
cdef inline tm localtime() nogil:
```

**Comment by jbrockmendel** [comment]
Any chance the CI failures are unrelated?

**Comment by scoder** [comment]
They look related to me, e.g.:
```
     File "<doctest time_pxd.test_localtime[1]>", line 1, in <module>
        assert ltp.tm_year == ltc.tm_year
    AttributeError: 'dict' object has no attribute 'tm_year'
    AttributeError: 'dict' object has no attribute 'tm_mon'
    AttributeError: 'dict' object has no attribute 'tm_wday'".
    …
```

**Comment by jbrockmendel** [comment]
> Let doctest do its work rather than calling "assert".

Trouble here is we don't get useful information when the test fails

**Comment by scoder** [comment]
> Trouble here is we don't get useful information when the test fails

There's nothing less informative than a bare `AssertionError`, though. What we could do is to print the difference of the two, which should be `0`. Or, rather, what I often do is
`>>> x == y  or  (x, y)`
to print the values if the assertion fails.

**Comment by scoder** [comment]
It is a bit unfortunate that there is an error case in `localtime()`. However unlikely it is, it now adds exception handling to the whole thing. :-/

**Comment by jbrockmendel** [comment]
Looks like this is still producing compilation errors

```
======================================================================
ERROR: runTest (__main__.CythonRunTestCase)
compiling (c/cy2) and running time_pxd
----------------------------------------------------------------------
Traceback (most recent call last):
  File "runtests.py", line 1289, in run
    ext_so_path = self.runCompileTest()
  File "runtests.py", line 958, in runCompileTest
    self.test_directory, self.expect_errors, self.expect_warnings, self.annotate)
  File "runtests.py", line 1223, in compile
    so_path = self.run_distutils(test_directory, module, workdir, incdir)
  File "runtests.py", line 1144, in run_distutils
    build_extension.run()
  File "/Users/travis/miniconda/lib/python2.7/distutils/command/build_ext.py", line 340, in run
    self.build_extensions()
  File "/Users/travis/miniconda/lib/python2.7/distutils/command/build_ext.py", line 449, in build_extensions
    self.build_extension(ext)
  File "runtests.py", line 520, in build_extension
    _build_ext.build_extension(self, ext)
  File "/Users/travis/miniconda/lib/python2.7/distutils/command/build_ext.py", line 499, in build_extension
    depends=ext.depends)
  File "/Users/travis/miniconda/lib/python2.7/distutils/ccompiler.py", line 574, in compile
    self._compile(obj, src, ext, cc_args, extra_postargs, pp_opts)
  File "/Users/travis/miniconda/lib/python2.7/distutils/unixccompiler.py", line 124, in _compile
    raise CompileError, msg
CompileError: command 'clang' failed with exit status 1
======================================================================
ERROR: runTest (__main__.CythonRunTestCase)
compiling (cpp/cy2) and running time_pxd
----------------------------------------------------------------------
Traceback (most recent call last):
  File "runtests.py", line 1289, in run
    ext_so_path = self.runCompileTest()
  File "runtests.py", line 958, in runCompileTest
    self.test_directory, self.expect_errors, self.expect_warnings, self.annotate)
  File "runtests.py", line 1223, in compile
    so_path = self.run_distutils(test_directory, module, workdir, incdir)
  File "runtests.py", line 1144, in run_distutils
    build_extension.run()
  File "/Users/travis/miniconda/lib/python2.7/distutils/command/build_ext.py", line 340, in run
    self.build_extensions()
  File "/Users/travis/miniconda/lib/python2.7/distutils/command/build_ext.py", line 449, in build_extensions
    self.build_extension(ext)
  File "runtests.py", line 520, in build_extension
    _build_ext.build_extension(self, ext)
  File "/Users/travis/miniconda/lib/python2.7/distutils/command/build_ext.py", line 499, in build_extension
    depends=ext.depends)
  File "/Users/travis/miniconda/lib/python2.7/distutils/ccompiler.py", line 574, in compile
    self._compile(obj, src, ext, cc_args, extra_postargs, pp_opts)
  File "/Users/travis/miniconda/lib/python2.7/distutils/unixccompiler.py", line 124, in _compile
    raise CompileError, msg
CompileError: command 'clang' failed with exit status 1
```

**Comment by scoder** [comment]
Those are test failures, because `tm_wday` also differs. Look at what CPython does:
https://github.com/python/cpython/blob/bb0b08540cc93e56f3f1bde1b39ce086d9e35fe1/Modules/timemodule.c#L397-L405

**Comment by scoder** [comment]
> Looks like this is still producing compilation errors

Ah, those are in the Py2.7 builds. Looks like `pytime.h` is a pure Py3 feature.

**Comment by jbrockmendel** [comment]
> Ah, those are in the Py2.7 builds. Looks like pytime.h is a pure Py3 feature.

So I guess disable the code/tests for >py35?

**Comment by jbrockmendel** [comment]
Rebased yesterday.  Looking at test logs, none of the failures look related to this PR (though the logs are huge, so worth double-checking)

Am I right in thinking that a/the barrier to releasing 3.0 is getting the CI passing?

**Comment by da-woods** [comment]
> Rebased yesterday. Looking at test logs, none of the failures look related to this PR (though the logs are huge, so worth double-checking)

I think you're right - the failures are all existing ones, not introduced by your PR

> Am I right in thinking that a/the barrier to releasing 3.0 is getting the CI passing?

I don't think it's the only barrier (I'm not sure what list of features is currently targetted for 3.0). I'm sure fixes to get the CI passing would be well-received though. I think the C++/pythran issues are a genuine bug (not sure if it's a bug in Pythran or Cython though...), the PyPy issues are a bug/omission in PyPy that's now been fixed (but the fixed version probably hasn't propagated to the CI yet), and the issues on older versions of Python 3 are probably just CI issues. The CI [may well be migrated in the near future though](https://github.com/cython/cython/pull/3943)

**Comment by robertwb** [comment]
The CI failures do look unrelated to this PR.
